File: u2flib-server-core/src/test/scala/com/yubico/webauthn/FinishAssertionSteps.scala
package com.yubico.webauthn


import java.security.PublicKey
import java.util.Optional

import com.yubico.scala.util.JavaConverters._
import com.yubico.u2f.crypto.Crypto
import com.yubico.u2f.data.messages.key.util.U2fB64Encoding
import com.yubico.webauthn.data.PublicKeyCredential
import com.yubico.webauthn.data.Base64UrlString
import com.yubico.webauthn.data.CollectedClientData
import com.yubico.webauthn.data.ArrayBuffer
import com.yubico.webauthn.data.AuthenticatorAssertionResponse

import scala.collection.JavaConverters._
import scala.util.Try

case class FinishAssertionSteps(
  request: PublicKeyCredentialRequestOptions,
  response: PublicKeyCredential[AuthenticatorAssertionResponse],
  callerTokenBindingId: Optional[Base64UrlString],
  origin: String,
  rpId: String,
  crypto: Crypto,
  credentialRepository: CredentialRepository,
) {

  sealed trait Step[A <: Step[_, _], B <: Step[_, _]] {
    protected def isFinished: Boolean = false
    protected def nextStep: B
    protected def result: Option[Boolean] = None
    protected def validate(): Unit

    private[webauthn] def next: Try[B] = validations map { _ => nextStep }
    private[webauthn] def prev: A
    private[webauthn] def validations: Try[Unit] = Try { validate() }

    def run: Try[Boolean] =
      if (isFinished) Try(result.get)
      else next flatMap { _.run }
  }

  private[webauthn] def begin: Step1 = Step1()
  def run: Try[Boolean] = begin.run

  case class Step1 private () extends Step[Step1, Step2] {
    override def prev = this
    override def nextStep = Step2(this)
    override def validate() = assert(_pubkey.isPresent, "Unknown credential ID.")

    private lazy val _pubkey: Optional[PublicKey] = (
      credentialRepository.lookup(response.id).asScala
        orElse credentialRepository.lookup(response.rawId).asScala
      ).asJava
    def pubkey: PublicKey = _pubkey.get
  }

  case class Step2 private (override val prev: Step1) extends Step[Step1, Step3] {
    override def validate() = {
      assert(response.response.clientData != null, "Missing client data.")
      assert(response.response.authenticatorData != null, "Missing authenticator data.")
      assert(response.response.signature != null, "Missing signature.")
    }
    override def nextStep = Step3(this)
  }

  case class Step3 private (override val prev: Step2) extends Step[Step2, Step4] {
    override def validate(): Unit = {
      assert(clientData != null, "Missing client data.")
    }
    override def nextStep = Step4(this)
    def clientData: CollectedClientData = response.response.collectedClientData
  }

  case class Step4 private (override val prev: Step3) extends Step[Step3, Step5] {
    override def validate() {
      assert(
        U2fB64Encoding.decode(response.response.collectedClientData.challenge).toVector == request.challenge,
        "Incorrect challenge."
      )
    }
    def nextStep = Step5(this)
  }

  case class Step5 private (override val prev: Step4) extends Step[Step4, Step6] {
    override def validate() {
      assert(
        response.response.collectedClientData.origin == origin,
        "Incorrect origin."
      )
    }
    override def nextStep = Step6(this)
  }

  case class Step6 private (override val prev: Step5) extends Step[Step5, Step7] {
    override def validate() {
      (callerTokenBindingId.asScala, response.response.collectedClientData.tokenBindingId.asScala) match {
        case (None, None) =>
        case (_, None) => throw new AssertionError("Token binding ID set by caller but not in attestation message.")
        case (None, _) => throw new AssertionError("Token binding ID set in attestation message but not by caller.")
        case (Some(callerToken), Some(responseToken)) =>
          assert(callerToken == responseToken, "Incorrect token binding ID.")
      }
    }
    override def nextStep = Step7(this)

    def clientDataJsonHash: ArrayBuffer = crypto.hash(response.response.clientDataJSON.toArray).toVector
  }
  case class Step7 private (override val prev: Step6) extends Step[Step6, Step8] {
    override def validate() {
      response.response.collectedClientData.clientExtensions.asScala foreach { extensions =>
        assert(
          request.extensions.isPresent,
          "Extensions were returned, but not requested."
        )

        assert(
          extensions.fieldNames.asScala.toSet subsetOf request.extensions.get.fieldNames.asScala.toSet,
          "Client extensions are not a subset of requested extensions."
        )
      }

      response.response.collectedClientData.authenticatorExtensions.asScala foreach { extensions =>
        assert(
          request.extensions.isPresent,
          "Extensions were returned, but not requested."
        )

        assert(
          extensions.fieldNames.asScala.toSet subsetOf request.extensions.get.fieldNames.asScala.toSet,
          "Authenticator extensions are not a subset of requested extensions."
        )
      }
    }
    override def nextStep = Step8(this)
  }

  case class Step8 private (override val prev: Step7) extends Step[Step7, Step9] {
    override def validate() {
      assert(
        response.response.parsedAuthenticatorData.rpIdHash == crypto.hash(rpId).toVector,
        "Wrong RP ID hash."
      )
    }
    override def nextStep = Step9(this)
  }

  case class Step9 private (override val prev: Step8) extends Step[Step8, Step10] {
    override def validate(): Unit = {
      val hashAlgorithm: String = response.response.collectedClientData.hashAlgorithm.toLowerCase
      assert(
        supportedHashAlgorithms map { _.toLowerCase } contains hashAlgorithm,
        s"Forbidden hash algorithm: ${hashAlgorithm}"
      )
    }
    override def nextStep = Step10(this)

    def clientDataJsonHash: ArrayBuffer = crypto.hash(response.response.clientDataJSON.toArray).toVector
    val supportedHashAlgorithms: List[String] = List("SHA-256")
  }

  case class Step10 (override val prev: Step9) extends Step[Step9, Finished] {
    override def validate() {
      assert(
        Try(
          crypto.checkSignature(
            prev.prev.prev.prev.prev.prev.prev.prev.prev.pubkey,
            signedBytes.toArray,
            response.response.signature.toArray,
          )
        ).isSuccess,

        "Invalid assertion signature."
      )
    }
    override def nextStep = Finished(this)

    val signedBytes: ArrayBuffer = response.response.authenticatorData ++ prev.clientDataJsonHash
  }

  case class Finished private (override val prev: Step10) extends Step[Step10, Finished] {
    override def validate() { /* No-op */ }
    override def isFinished = true
    override def nextStep = this

    val success: Boolean = true
  }

}

File: webauthn-server-core/src/main/java/com/yubico/webauthn/impl/FinishAssertionSteps.java
package com.yubico.webauthn.impl;


import com.yubico.u2f.exceptions.U2fBadInputException;
import com.yubico.webauthn.data.ByteArray;
import com.yubico.webauthn.CredentialRepository;
import com.yubico.webauthn.Crypto;
import com.yubico.webauthn.data.AssertionRequest;
import com.yubico.webauthn.data.AssertionResult;
import com.yubico.webauthn.data.AuthenticatorAssertionResponse;
import com.yubico.webauthn.data.CollectedClientData;
import com.yubico.webauthn.data.PublicKeyCredential;
import com.yubico.webauthn.data.RegisteredCredential;
import com.yubico.webauthn.data.UserVerificationRequirement;
import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedList;
import java.util.List;
import java.util.Optional;
import lombok.Builder;
import lombok.Value;
import lombok.extern.slf4j.Slf4j;

import static com.yubico.util.ExceptionUtil.assure;


@Builder
@Slf4j
public class FinishAssertionSteps {

    private static final String CLIENT_DATA_TYPE = "webauthn.get";

    private final AssertionRequest request;
    private final PublicKeyCredential<AuthenticatorAssertionResponse> response;
    private final Optional<ByteArray> callerTokenBindingId;
    private final List<String> origins;
    private final String rpId;
    private final Crypto crypto;
    private final CredentialRepository credentialRepository;

    @Builder.Default
    private final boolean allowMissingTokenBinding = false;
    @Builder.Default
    private final boolean validateTypeAttribute = true;
    @Builder.Default
    private final boolean validateSignatureCounter = true;
    @Builder.Default
    private final boolean allowUnrequestedExtensions = false;

    public Step0 begin() {
        return new Step0();
    }

    public AssertionResult run() {
        return begin().run();
    }

    private interface Step<A extends Step<?, ?>, B extends Step<?, ?>> {
        B nextStep();

        void validate();

        List<String> getPrevWarnings();

        default boolean isFinished() {
            return false;
        }

        default Optional<AssertionResult> result() {
            return Optional.empty();
        }

        default List<String> getWarnings() {
            return Collections.emptyList();
        }

        default List<String> allWarnings() {
            List<String> result = new ArrayList<>(getPrevWarnings().size() + getWarnings().size());
            result.addAll(getPrevWarnings());
            result.addAll(getWarnings());
            return Collections.unmodifiableList(result);
        }

        default B next() {
            validate();
            return nextStep();
        }

        default AssertionResult run() {
            if (isFinished()) {
                return result().get();
            } else {
                return next().run();
            }
        }
    }

    @Value
    public class Step0 implements Step<Step0, Step1> {
        @Override
        public Step1 nextStep() {
            return new Step1(username().get(), userHandle().get(), allWarnings());
        }

        @Override
        public void validate() {
            assure(
                request.getUsername().isPresent() || response.getResponse().getUserHandle().isPresent(),
                "At least one of username and user handle must be given; none was."
            );
            assure(
                userHandle().isPresent(),
                "No user found for username: %s, userHandle: %s",
                request.getUsername(), response.getResponse().getUserHandle()
            );
            assure(
                username().isPresent(),
                "No user found for username: %s, userHandle: %s",
                request.getUsername(), response.getResponse().getUserHandle()
            );
        }

        @Override
        public List<String> getPrevWarnings() {
            return Collections.emptyList();
        }

        private Optional<ByteArray> userHandle() {
            return response.getResponse().getUserHandle()
                .map(Optional::of)
                .orElseGet(() -> credentialRepository.getUserHandleForUsername(request.getUsername().get()));
        }

        private Optional<String> username() {
            return request.getUsername()
                .map(Optional::of)
                .orElseGet(() -> credentialRepository.getUsernameForUserHandle(response.getResponse().getUserHandle().get()));
        }
    }

    @Value
    public class Step1 implements Step<Step0, Step2> {
        private final String username;
        private final ByteArray userHandle;
        private final List<String> prevWarnings;

        @Override
        public Step2 nextStep() {
            return new Step2(username, userHandle, allWarnings());
        }

        @Override
        public void validate() {
            request.getPublicKeyCredentialRequestOptions().getAllowCredentials().ifPresent(allowed -> {
                assure(
                    allowed.stream().anyMatch(allow -> allow.getId().equals(response.getId())),
                    "Unrequested credential ID: %s",
                    response.getId()
                );
            });
        }
    }

    @Value
    public class Step2 implements Step<Step1, Step3> {
        private final String username;
        private final ByteArray userHandle;
        private final List<String> prevWarnings;

        @Override
        public Step3 nextStep() {
            return new Step3(username, userHandle, allWarnings());
        }

        @Override
        public void validate() {
            Optional<RegisteredCredential> registration = credentialRepository.lookup(response.getId(), userHandle);

            assure(
                registration.isPresent(),
                "Unknown credential: %s",
                response.getId()
            );

            assure(
                userHandle.equals(registration.get().getUserHandle()),
                "User handle %s does not own credential %s",
                userHandle, response.getId()
            );
        }
    }

    @Value
    public class Step3 implements Step<Step2, Step4> {
        private final String username;
        private final ByteArray userHandle;
        private final List<String> prevWarnings;

        @Override
        public Step4 nextStep() {
            return new Step4(username, userHandle, credential(), allWarnings());
        }

        @Override
        public void validate() {
            assure(
                maybeCredential().isPresent(),
                "Unknown credential. Credential ID: %s, user handle: %s",
                response.getId(), userHandle
            );
        }

        private Optional<RegisteredCredential> maybeCredential() {
            return credentialRepository.lookup(response.getId(), userHandle);
        }

        public RegisteredCredential credential() {
            return maybeCredential().get();
        }
    }

    @Value
    public class Step4 implements Step<Step3, Step5> {

        private final String username;
        private final ByteArray userHandle;
        private final RegisteredCredential credential;
        private final List<String> prevWarnings;

        @Override
        public void validate() {
            assure(clientData().getBytes() != null, "Missing client data.");
            assure(authenticatorData().getBytes() != null, "Missing authenticator data.");
            assure(signature().getBytes() != null, "Missing signature.");
        }

        @Override
        public Step5 nextStep() {
            return new Step5(username, userHandle, credential, allWarnings());
        }

        public ByteArray authenticatorData() {
            return response.getResponse().getAuthenticatorData();
        }

        public ByteArray clientData() {
            return response.getResponse().getClientDataJSON();
        }

        public ByteArray signature() {
            return response.getResponse().getSignature();
        }
    }

    @Value
    public class Step5 implements Step<Step4, Step6> {
        private final String username;
        private final ByteArray userHandle;
        private final RegisteredCredential credential;
        private final List<String> prevWarnings;

        // Nothing to do
        @Override
        public void validate() {
        }

        @Override
        public Step6 nextStep() {
            return new Step6(username, userHandle, credential, allWarnings());
        }
    }

    @Value
    public class Step6 implements Step<Step5, Step7> {
        private final String username;
        private final ByteArray userHandle;
        private final RegisteredCredential credential;
        private final List<String> prevWarnings;

        @Override
        public void validate() {
            assure(clientData() != null, "Missing client data.");
        }

        @Override
        public Step7 nextStep() {
            return new Step7(username, userHandle, credential, clientData(), allWarnings());
        }

        public CollectedClientData clientData() {
            return response.getResponse().getClientData();
        }
    }

    @Value
    public class Step7 implements Step<Step6, Step8> {

        private final String username;
        private final ByteArray userHandle;
        private final RegisteredCredential credential;
        private final CollectedClientData clientData;
        private final List<String> prevWarnings;

        private List<String> warnings = new LinkedList<>();

        @Override
        public List<String> getWarnings() {
            return Collections.unmodifiableList(warnings);
        }

        @Override
        public void validate() {
            if (!
                CLIENT_DATA_TYPE.equals(clientData.getType())
            ) {
                final String message = String.format(
                    "The \"type\" in the client data must be exactly \"%s\", was: %s", CLIENT_DATA_TYPE, clientData.getType()
                );
                if (validateTypeAttribute) {
                    throw new IllegalArgumentException(message);
                } else {
                    warnings.add(message);
                }
            }
        }

        @Override
        public Step8 nextStep() {
            return new Step8(username, userHandle, credential, allWarnings());
        }
    }

    @Value
    public class Step8 implements Step<Step7, Step9> {
        private final String username;
        private final ByteArray userHandle;
        private final RegisteredCredential credential;
        private final List<String> prevWarnings;

        @Override
        public void validate() {
            assure(
                request.getPublicKeyCredentialRequestOptions().getChallenge().equals(response.getResponse().getClientData().getChallenge()),
                "Incorrect challenge."
            );
        }

        @Override
        public Step9 nextStep() {
            return new Step9(username, userHandle, credential, allWarnings());
        }
    }

    @Value
    public class Step9 implements Step<Step8, Step10> {
        private final String username;
        private final ByteArray userHandle;
        private final RegisteredCredential credential;
        private final List<String> prevWarnings;

        @Override
        public void validate() {
            final String responseOrigin;
            responseOrigin = response.getResponse().getClientData().getOrigin();

            if (origins.stream().noneMatch(o -> o.equals(responseOrigin))) {
                throw new IllegalArgumentException("Incorrect origin: " + responseOrigin);
            }
        }

        @Override
        public Step10 nextStep() {
            return new Step10(username, userHandle, credential, allWarnings());
        }
    }

    @Value
    public class Step10 implements Step<Step9, Step11> {
        private final String username;
        private final ByteArray userHandle;
        private final RegisteredCredential credential;
        private final List<String> prevWarnings;

        @Override
        public void validate() {
            TokenBindingValidator.validate(response.getResponse().getClientData().getTokenBinding(), callerTokenBindingId);
        }

        @Override
        public Step11 nextStep() {
            return new Step11(username, userHandle, credential, allWarnings());
        }
    }

    @Value
    public class Step11 implements Step<Step10, Step12> {
        private final String username;
        private final ByteArray userHandle;
        private final RegisteredCredential credential;
        private final List<String> prevWarnings;

        @Override
        public void validate() {
            assure(
                new ByteArray(crypto.hash(rpId)).equals(response.getResponse().getParsedAuthenticatorData().getRpIdHash()),
                "Wrong RP ID hash."
            );
        }

        @Override
        public Step12 nextStep() {
            return new Step12(username, userHandle, credential, allWarnings());
        }
    }

    @Value
    public class Step12 implements Step<Step11, Step13> {
        private final String username;
        private final ByteArray userHandle;
        private final RegisteredCredential credential;
        private final List<String> prevWarnings;

        @Override
        public void validate() {
            if (request.getPublicKeyCredentialRequestOptions().getUserVerification() == UserVerificationRequirement.REQUIRED) {
                assure(
                    response.getResponse().getParsedAuthenticatorData().getFlags().UV,
                    "User Verification is required."
                );
            }
        }

        @Override
        public Step13 nextStep() {
            return new Step13(username, userHandle, credential, allWarnings());
        }
    }

    @Value
    public class Step13 implements Step<Step12, Step14> {
        private final String username;
        private final ByteArray userHandle;
        private final RegisteredCredential credential;
        private final List<String> prevWarnings;

        @Override
        public void validate() {
            if (request.getPublicKeyCredentialRequestOptions().getUserVerification() != UserVerificationRequirement.REQUIRED) {
                assure(
                    response.getResponse().getParsedAuthenticatorData().getFlags().UP,
                    "User Presence is required."
                );
            }
        }

        @Override
        public Step14 nextStep() {
            return new Step14(username, userHandle, credential, allWarnings());
        }
    }

    @Value
    public class Step14 implements Step<Step13, Step15> {
        private final String username;
        private final ByteArray userHandle;
        private final RegisteredCredential credential;
        private final List<String> prevWarnings;

        @Override
        public void validate() {
            if (!allowUnrequestedExtensions) {
                ExtensionsValidation.validate(request.getPublicKeyCredentialRequestOptions().getExtensions(), response);
            }
        }

        @Override
        public List<String> getWarnings() {
            try {
                ExtensionsValidation.validate(request.getPublicKeyCredentialRequestOptions().getExtensions(), response);
                return Collections.emptyList();
            } catch (Exception e) {
                return Collections.unmodifiableList(Collections.singletonList(e.getMessage()));
            }
        }

        @Override
        public Step15 nextStep() {
            return new Step15(username, userHandle, credential, allWarnings());
        }
    }

    @Value
    public class Step15 implements Step<Step14, Step16> {
        private final String username;
        private final ByteArray userHandle;
        private final RegisteredCredential credential;
        private final List<String> prevWarnings;

        @Override
        public void validate() {
            assure(clientDataJsonHash() != null, "Failed to compute hash of client data");
        }

        @Override
        public Step16 nextStep() {
            return new Step16(username, userHandle, credential, clientDataJsonHash(), allWarnings());
        }

        public ByteArray clientDataJsonHash() {
            return new ByteArray(crypto.hash(response.getResponse().getClientDataJSON().getBytes()));
        }
    }

    @Value
    public class Step16 implements Step<Step15, Step17> {
        private final String username;
        private final ByteArray userHandle;
        private final RegisteredCredential credential;
        private final ByteArray clientDataJsonHash;
        private final List<String> prevWarnings;

        @Override
        public void validate() {
            if (!
                crypto.verifySignature(
                    credential.publicKey,
                    signedBytes().getBytes(),
                    response.getResponse().getSignature().getBytes()
                )
            ) {
                throw new IllegalArgumentException("Invalid assertion signature.");
            }
        }

        @Override
        public Step17 nextStep() {
            return new Step17(username, userHandle, allWarnings());
        }

        public ByteArray signedBytes() {
            return response.getResponse().getAuthenticatorData().concat(clientDataJsonHash);
        }
    }

    @Value
    public class Step17 implements Step<Step16, Finished> {
        private final String username;
        private final ByteArray userHandle;
        private final List<String> prevWarnings;

        @Override
        public void validate() {
            if (validateSignatureCounter) {
                assure(
                    signatureCounterValid(),
                    "Signature counter must increase. Stored value: %s, received value: %s",
                    storedSignatureCountBefore(), assertionSignatureCount()
                );
            }
        }

        private boolean signatureCounterValid() {
            return assertionSignatureCount() == 0
                || assertionSignatureCount() > storedSignatureCountBefore();
        }

        @Override
        public Finished nextStep() {
            return new Finished(username, userHandle, assertionSignatureCount(), signatureCounterValid(), allWarnings());
        }

        private long storedSignatureCountBefore() {
            return credentialRepository.lookup(response.getId(), userHandle)
                .map(RegisteredCredential::getSignatureCount)
                .orElse(0L);
        }

        private long assertionSignatureCount() {
            return response.getResponse().getParsedAuthenticatorData().getSignatureCounter();
        }
    }

    @Value
    public class Finished implements Step<Step17, Finished> {
        private final String username;
        private final ByteArray userHandle;
        private final long assertionSignatureCount;
        private final boolean signatureCounterValid;
        private final List<String> prevWarnings;

        @Override
        public void validate() { /* No-op */ }

        @Override
        public boolean isFinished() {
            return true;
        }

        @Override
        public Finished nextStep() {
            return this;
        }

        @Override
        public Optional<AssertionResult> result() {
            return Optional.of(AssertionResult.builder()
                .credentialId(response.getId())
                .signatureCount(assertionSignatureCount)
                .signatureCounterValid(signatureCounterValid)
                .success(true)
                .username(username)
                .userHandle(userHandle)
                .warnings(allWarnings())
                .build()
            );
        }

    }

}
